src/dlio/odom.cc

```diff
@@ -447,6 +449,53 @@ void dlio::OdomNode::publishToROS(pcl::PointCloud<PointType>::ConstPtr published
   br.sendTransform(transformStamped);
 }
 
+// Function for correcting the pointcloud tilestamp to more accurate represent its pose
+ros::Time dlio::OdomNode::correctTimestamp(ros::Time original_stamp) {
+  
+  // Pose buffer is empty, return previous stamp
+  if (this->pose_buffer.empty()) {
+    printf("Correct timestamp failed 1\n");
+    return original_stamp;
+  }
+
+  // Required variables
+  auto curr_pose = this->pose_buffer.begin(); // Current pose
+  auto prev_pose = this->pose_buffer.begin(); // Previous pose
+  auto best_prev_pose = this->pose_buffer.begin(); // Best previous pose
+  auto best_curr_pose = this->pose_buffer.begin(); // Best current pose
+  double best_cost = -1;
+
+  // Loop through buffer
+  while (curr_pose->second.toSec() > original_stamp.toSec()) {
+    
+    // End of buffer reached but scan time not reached yet, return previous stamp (end-1 is last valid entry)
+    if (curr_pose == this->pose_buffer.end()-1) {
+      printf("Correct timestamp failed 2\n");
+      return original_stamp;
+    }
+
+    // Update iterators
+    prev_pose = curr_pose;
+    curr_pose++;
+    
+    // Find the sum of distances between lidarPose and 2 consecutive poses
+    double cost = (prev_pose->first - this->lidarPose.p).norm() + (curr_pose->first - this->lidarPose.p).norm();
+    
+    // Save cost and pose if better than previous best
+    if (best_cost == -1 || best_cost > cost) {
+      best_cost = cost;
+      best_prev_pose = prev_pose;
+      best_curr_pose = curr_pose;
+    }
```

Comment:
This assumption is correct if the closest pose is behind the `lidarPose`. If the pose is in front of `lidarPose`, the previous best pose will be the one before that, and thus be further than the next pose. You would need to catch at which point you would go from behind the `lidarPose` to in front of it, for instance by calculating the angle in the frame of the `lidarPose` between the pose and the coordinate system of the lidar pose.